File: internal/webserver/projects.go
package webserver

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"path/filepath"

	"github.com/RomkaLTU/trau/internal/hubstore"
	"github.com/RomkaLTU/trau/internal/logger"
	"github.com/RomkaLTU/trau/internal/registry"
)

// ProjectView is one project as the hub serves it: its identifier, display name,
// and member repo roots in order. It carries roots rather than repos so the repos
// resource stays the single source of truth for what each root is.
type ProjectView struct {
	ID    string   `json:"id"`
	Name  string   `json:"name"`
	Repos []string `json:"repos"`
}

// ProjectsResponse is the /api/v1/projects resource: every project the hub holds,
// ordered by display name.
type ProjectsResponse struct {
	Projects []ProjectView `json:"projects"`
}

// ProjectRequest is the body of the project create and rename calls.
type ProjectRequest struct {
	Name string `json:"name"`
}

// ProjectRepoRequest is the body of POST /api/v1/projects/{project}/repos: an
// already-registered repo addressed by name or root.
type ProjectRepoRequest struct {
	Repo string `json:"repo"`
}

// ProjectRemovalRepo is one member a project-wide removal reached, with the reason
// the hub refused it on a member that stayed.
type ProjectRemovalRepo struct {
	Name   string `json:"name"`
	Root   string `json:"root"`
	Reason string `json:"reason,omitempty"`
}

// ProjectRemoval is the outcome of DELETE /api/v1/projects/{project}?forget=1.
// Project is the pre-removal snapshot, so a caller can name what it acted on even
// once the row is gone. Removed and Blocked ride back together because the removal
// is per member rather than all-or-nothing, and ProjectDeleted stays false while
// any member is blocked so those members keep their group.
type ProjectRemoval struct {
	Project        ProjectView          `json:"project"`
	Removed        []ProjectRemovalRepo `json:"removed"`
	Blocked        []ProjectRemovalRepo `json:"blocked"`
	ProjectDeleted bool                 `json:"project_deleted"`
}

// handleProjects lists the projects (GET) or creates one (POST).
func (s *Server) handleProjects(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		s.adoptRepoTrackers()
		projects, err := s.stores.Projects().List()
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "failed to list projects: " + err.Error()})
			return
		}
		writeJSON(w, http.StatusOK, ProjectsResponse{Projects: projectViews(projects)})
	case http.MethodPost:
		s.createProject(w, r)
	default:
		w.Header().Set("Allow", "GET, POST")
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
	}
}

// handleProject renames (PATCH) or deletes (DELETE) one project. The delete splits
// the way a repo row's does: bare drops the grouping only, ?forget=1 takes every
// member repo off the hub with it.
func (s *Server) handleProject(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodPatch, http.MethodPut:
		s.renameProject(w, r)
	case http.MethodDelete:
		if r.URL.Query().Get("forget") == "1" {
			s.forgetProjectRepos(w, r)
			return
		}
		s.deleteProject(w, r)
	default:
		w.Header().Set("Allow", "PATCH, DELETE")
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
	}
}

// createProject files a new, empty project under the requested name. Grouping
// cannot widen the startable set, but the write follows the registration gate all
// the same, keeping every write that reshapes the repo list under one rule.
func (s *Server) createProject(w http.ResponseWriter, r *http.Request) {
	if s.denyRegistrationIfExposed(w, "creating a project") {
		return
	}
	var req ProjectRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
		return
	}
	proj, err := s.stores.Projects().Create(req.Name)
	if err != nil {
		writeProjectError(w, err, "failed to create project")
		return
	}
	writeJSON(w, http.StatusCreated, projectView(proj))
}

func (s *Server) renameProject(w http.ResponseWriter, r *http.Request) {
	if s.denyRegistrationIfExposed(w, "renaming a project") {
		return
	}
	var req ProjectRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
		return
	}
	proj, err := s.stores.Projects().Rename(r.PathValue("project"), req.Name)
	if err != nil {
		writeProjectError(w, err, "failed to rename project")
		return
	}
	writeJSON(w, http.StatusOK, projectView(proj))
}

// deleteProject drops the grouping only; its member repos stay registered.
func (s *Server) deleteProject(w http.ResponseWriter, r *http.Request) {
	if s.denyRegistrationIfExposed(w, "deleting a project") {
		return
	}
	id := r.PathValue("project")
	proj, err := s.stores.Projects().Get(id)
	if err != nil {
		writeProjectError(w, err, "failed to read project")
		return
	}
	if err := s.stores.Projects().Delete(id); err != nil {
		writeProjectError(w, err, "failed to delete project")
		return
	}
	writeJSON(w, http.StatusOK, projectView(proj))
}

// forgetProjectRepos takes every member repo off the hub in one call, so the folder
// that registered them clears in one action instead of one confirm per repo. Each
// member gets forgetRepo's treatment — registration, known-repos row, cached
// attachments and tracker sync state — and nothing on disk is touched. It is
// deliberately partial rather than all-or-nothing: a member the hub refuses would
// otherwise leave the project permanently un-clearable, so it stays registered
// while its siblings go, and keeps the project row with it. It follows the same
// exposure gate as registration.
func (s *Server) forgetProjectRepos(w http.ResponseWriter, r *http.Request) {
	if s.denyRegistrationIfExposed(w, "removing a project's repos") {
		return
	}
	id := r.PathValue("project")
	proj, err := s.stores.Projects().Get(id)
	if err != nil {
		writeProjectError(w, err, "failed to read project")
		return
	}
	out := ProjectRemoval{
		Project: projectView(proj),
		Removed: []ProjectRemovalRepo{},
		Blocked: []ProjectRemovalRepo{},
	}
	for _, root := range proj.Repos {
		repo, listed := s.matchListedRepo(root)
		if !listed {
			// A membership the repos list no longer answers for has nothing left to
			// remove, so dropping the row is the whole removal.
			if err := s.stores.Projects().ForgetRoot(root); err != nil {
				logger.Verbosef("drop stale project membership for %s: %v", root, err)
			}
			out.Removed = append(out.Removed, ProjectRemovalRepo{Name: filepath.Base(root), Root: root})
			continue
		}
		if reason := s.removalRefusal(repo); reason != "" {
			out.Blocked = append(out.Blocked, ProjectRemovalRepo{Name: repo.Name, Root: repo.Root, Reason: reason})
			continue
		}
		if err := s.forgetMemberRepo(repo.Root); err != nil {
			out.Blocked = append(out.Blocked, ProjectRemovalRepo{Name: repo.Name, Root: repo.Root, Reason: err.Error()})
			continue
		}
		out.Removed = append(out.Removed, ProjectRemovalRepo{Name: repo.Name, Root: repo.Root})
	}
	if len(out.Blocked) == 0 {
		// The last membership to go prunes the project row with it, so a missing row
		// here is the delete already done rather than a failure.
		if err := s.stores.Projects().Delete(id); err != nil && !errors.Is(err, hubstore.ErrProjectNotFound) {
			writeProjectError(w, err, "failed to delete project")
			return
		}
		out.ProjectDeleted = true
	}
	writeJSON(w, http.StatusOK, out)
}

// removalRefusal is why the hub keeps a member a project-wide removal aimed at,
// empty when the removal goes through. Both conflicts forgetRepo answers with a 409
// land here as a per-member reason instead, since one un-removable member must not
// strand its siblings.
func (s *Server) removalRefusal(repo registry.Repo) string {
	if _, ok := matchRoot(s.workspace, repo.Root); ok {
		return seededRepoRefusal(repo.Name, "removed")
	}
	if s.repoIsLive(repo.Root) {
		return liveLoopRefusal(repo.Name)
	}
	return ""
}

// forgetMemberRepo removes one member from the hub, mirroring forgetRepo. Only the
// registration drop can leave the repo listed, so it is the one failure the caller
// records against the member; the rest is best-effort cleanup.
func (s *Server) forgetMemberRepo(root string) error {
	if err := s.stores.Registrations().Forget(root); err != nil {
		return fmt.Errorf("failed to remove repo: %w", err)
	}
	if err := s.stores.Projects().ForgetRoot(root); err != nil {
		logger.Verbosef("drop project membership for %s: %v", root, err)
	}
	s.dropUnregisteredRepoState(root)
	return nil
}

// handleProjectRepos adds an already-registered repo to a project (POST), moving
// it out of whichever project held it.
func (s *Server) handleProjectRepos(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.Header().Set("Allow", http.MethodPost)
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
		return
	}
	if s.denyRegistrationIfExposed(w, "adding a repo to a project") {
		return
	}
	var req ProjectRepoRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
		return
	}
	repo, ok := s.findRepo(req.Repo)
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": fmt.Sprintf("repo %q is not known to the hub", req.Repo)})
		return
	}
	proj, err := s.stores.Projects().AddRepo(r.PathValue("project"), repo.Root)
	if err != nil {
		writeProjectError(w, err, "failed to add repo to project")
		return
	}
	// The stored tracker, not projectTracker: a project being assembled repo by
	// repo is momentarily single-member, and adopting there would claim the first
	// joiner's own keys as the project default.
	keys, err := s.stores.Projects().Tracker(proj.ID)
	if err == nil {
		err = s.seedRepoTracker(repo.Root, keys)
	}
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{
			"error": fmt.Sprintf("repo %q joined the project but seeding its tracker failed: %v", repo.Name, err),
		})
		return
	}
	writeJSON(w, http.StatusOK, projectView(proj))
}

// handleProjectRepo removes a repo from a project (DELETE), leaving it registered
// and standalone.
func (s *Server) handleProjectRepo(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodDelete {
		w.Header().Set("Allow", http.MethodDelete)
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"error": "method not allowed"})
		return
	}
	if s.denyRegistrationIfExposed(w, "removing a repo from a project") {
		return
	}
	ident := r.PathValue("repo")
	repo, ok := s.findRepo(ident)
	if !ok {
		writeJSON(w, http.StatusNotFound, map[string]string{"error": fmt.Sprintf("repo %q is not known to the hub", ident)})
		return
	}
	proj, err := s.stores.Projects().RemoveRepo(r.PathValue("project"), repo.Root)
	if err != nil {
		writeProjectError(w, err, "failed to remove repo from project")
		return
	}
	writeJSON(w, http.StatusOK, projectView(proj))
}

// writeProjectError maps the store's sentinels onto their status codes.
func writeProjectError(w http.ResponseWriter, err error, action string) {
	switch {
	case errors.Is(err, hubstore.ErrProjectNotFound), errors.Is(err, hubstore.ErrProjectRepoNotFound):
		writeJSON(w, http.StatusNotFound, map[string]string{"error": err.Error()})
	case errors.Is(err, hubstore.ErrProjectNameEmpty):
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
	default:
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": action + ": " + err.Error()})
	}
}

func projectViews(projects []hubstore.Project) []ProjectView {
	views := make([]ProjectView, 0, len(projects))
	for _, proj := range projects {
		views = append(views, projectView(proj))
	}
	return views
}

func projectView(proj hubstore.Project) ProjectView {
	repos := proj.Repos
	if repos == nil {
		repos = []string{}
	}
	return ProjectView{ID: proj.ID, Name: proj.Name, Repos: repos}
}

File: web/src/lib/projects.ts
import { queryOptions, useQuery } from '@tanstack/react-query'

import { apiFetch } from './api'
import { removeBlocked, type RepoView } from './instances'
import { matchesTerms } from './settings'

// ProjectView is one logical project: a group of registered repos under a display
// name. Members are repo roots, joined against the repos list by the caller.
export interface ProjectView {
  id: string
  name: string
  repos: string[]
}

export interface ProjectsResponse {
  projects: ProjectView[]
}

async function fetchProjects(): Promise<ProjectsResponse> {
  const res = await apiFetch('/api/v1/projects')
  if (!res.ok) {
    throw new Error(`projects request failed: ${res.status}`)
  }
  return res.json()
}

// projectForRoot names the project holding a repo root, empty when none does.
export async function projectForRoot(root: string): Promise<string> {
  const { projects } = await fetchProjects()
  return projects.find((p) => p.repos.includes(root))?.id ?? ''
}

export const projectsQueryOptions = queryOptions({
  queryKey: ['projects'],
  queryFn: fetchProjects,
  refetchInterval: 5000,
})

// projectNameForRoot names the project holding a repo root, empty when none
// does. A single-member project never forms a group, so its name has nowhere to
// surface but the repo's own row.
export function projectNameForRoot(
  root: string,
  projects: readonly ProjectView[],
): string {
  return projects.find((p) => p.repos.includes(root))?.name ?? ''
}

// RepoRow is one row of a repo list — the switcher's, or the instances page's. A
// project holding two or more listed repos renders as a named group; everything
// else — a single-member project, a repo no project holds — renders as a bare repo.
export interface RepoRow {
  project: ProjectView | null
  repos: RepoView[]
}

// groupRepos folds the projects into an already-ordered repo list. A group takes
// the place of its highest-ranked member, so the caller's ordering — live loops
// first, then most recently used — carries over unchanged.
export function groupRepos(
  ordered: readonly RepoView[],
  projects: readonly ProjectView[],
): RepoRow[] {
  const views = new Map(ordered.map((repo) => [repo.root, repo]))
  const owner = new Map<string, ProjectView>()
  for (const project of projects) {
    const listed = project.repos.filter((root) => views.has(root))
    if (listed.length < 2) continue
    for (const root of listed) owner.set(root, project)
  }
  const rows: RepoRow[] = []
  const grouped = new Set<string>()
  for (const repo of ordered) {
    const project = owner.get(repo.root)
    if (!project) {
      rows.push({ project: null, repos: [repo] })
      continue
    }
    if (grouped.has(project.id)) continue
    grouped.add(project.id)
    rows.push({
      project,
      repos: project.repos.flatMap((root) => views.get(root) ?? []),
    })
  }
  return rows
}

// filterRepoRows narrows the switcher to a query, matched case-insensitively
// against a repo's name or its root. A group whose own name matches keeps every
// member: naming the project reads as picking it, not as sifting it. A blank query
// hands the same rows back, so memoized rows stay put.
export function filterRepoRows(rows: RepoRow[], query: string): RepoRow[] {
  const needle = query.trim().toLowerCase()
  if (needle === '') return rows
  const kept: RepoRow[] = []
  for (const row of rows) {
    if (row.project?.name.toLowerCase().includes(needle)) {
      kept.push(row)
      continue
    }
    const repos = row.repos.filter(
      (repo) =>
        repo.name.toLowerCase().includes(needle) ||
        repo.root.toLowerCase().includes(needle),
    )
    if (repos.length > 0) kept.push({ ...row, repos })
  }
  return kept
}

// repoQualifiers tells same-named repos apart on the line that names them: a name
// only one repo carries is left alone, a name two repos share gains the project
// holding it, and a name their projects cannot tell apart falls back to an
// abbreviated root. Keyed by root, which is unique per hub.
export function repoQualifiers(
  repos: readonly RepoView[],
  projects: readonly ProjectView[],
): Map<string, string> {
  const byName = new Map<string, RepoView[]>()
  for (const repo of repos) {
    const shared = byName.get(repo.name) ?? []
    shared.push(repo)
    byName.set(repo.name, shared)
  }
  const qualifiers = new Map<string, string>()
  for (const shared of byName.values()) {
    if (shared.length < 2) continue
    const byProject = new Map<string, RepoView[]>()
    for (const repo of shared) {
      const name = projectNameForRoot(repo.root, projects)
      const holders = byProject.get(name) ?? []
      holders.push(repo)
      byProject.set(name, holders)
    }
    for (const [project, holders] of byProject) {
      if (project !== '' && holders.length === 1) {
        qualifiers.set(holders[0].root, project)
        continue
      }
      for (const [root, tail] of abbreviateRoots(holders.map((r) => r.root))) {
        qualifiers.set(root, tail)
      }
    }
  }
  return qualifiers
}

// abbreviateRoots is the shortest tail of each root no other root in the set
// ends with, so a row spends only as much of its line on a path as it takes to
// be tellable apart.
function abbreviateRoots(roots: readonly string[]): Map<string, string> {
  const parts = roots.map((root) => root.split('/').filter(Boolean))
  const deepest = Math.max(...parts.map((p) => p.length))
  for (let take = 2; take < deepest; take++) {
    const tails = parts.map((p) => pathTail(p, take))
    if (new Set(tails).size === roots.length) {
      return new Map(roots.map((root, i) => [root, tails[i]]))
    }
  }
  return new Map(roots.map((root) => [root, root]))
}

function pathTail(parts: readonly string[], take: number): string {
  if (parts.length <= take) return `/${parts.join('/')}`
  return `…/${parts.slice(-take).join('/')}`
}

// expandedOnOpen is the switcher's opening disclosure: every project row shut
// but the one holding the scoped repo, so the list stays short without hiding
// where the work currently is. Nothing is persisted — closing the dialog drops
// the state and the next open starts here again.
export function expandedOnOpen(
  rows: readonly RepoRow[],
  repo: string,
): Set<string> {
  const holder = rows.find(
    (row) => row.project !== null && row.repos.some((r) => r.name === repo),
  )
  return new Set(holder?.project ? [holder.project.id] : [])
}

export function toggleExpanded(
  expanded: ReadonlySet<string>,
  id: string,
): Set<string> {
  const next = new Set(expanded)
  if (!next.delete(id)) next.add(id)
  return next
}

// projectAnchor names the repo a project's shared tracker surfaces read from:
// its first listed member. Every member of a project talks to the same tracker,
// so anchoring gives the project one inbox instead of one per repo. A repo no
// project holds — and a single-member project — anchors on itself.
export function projectAnchor(
  repo: string,
  repos: readonly RepoView[],
  projects: readonly ProjectView[],
): string {
  const root = repos.find((r) => r.name === repo)?.root
  const project = root ? projects.find((p) => p.repos.includes(root)) : undefined
  if (!project) return repo
  const names = new Map(repos.map((r) => [r.root, r.name]))
  for (const member of project.repos) {
    const name = names.get(member)
    if (name) return name
  }
  return repo
}

// useProjectRepo resolves a scoped repo to its project's anchor, so the tracker
// inbox and its badge stay on one queue however the switcher moves between
// members.
export function useProjectRepo(repo: string, repos: readonly RepoView[]): string {
  const { data } = useQuery(projectsQueryOptions)
  return projectAnchor(repo, repos, data?.projects ?? [])
}

// ProjectMembers are the repos a project's work can be started in: its holder's
// identifier and every member the repos list knows, in project order. A repo no
// project holds resolves to none.
export interface ProjectMembers {
  project: string
  members: RepoView[]
}

export function projectMembers(
  repo: string,
  repos: readonly RepoView[],
  projects: readonly ProjectView[],
): ProjectMembers {
  const root = repos.find((r) => r.name === repo)?.root
  const project = root ? projects.find((p) => p.repos.includes(root)) : undefined
  if (!project) return { project: '', members: [] }
  const views = new Map(repos.map((r) => [r.root, r]))
  return {
    project: project.id,
    members: project.repos.flatMap((member) => views.get(member) ?? []),
  }
}

// StartRepoHint is the hub's answer to which member repo a ticket should start
// in: suggested names one of the project's members, reason says what earned it.
export interface StartRepoHint {
  project: string
  ticket?: string
  repos: { name: string; root: string }[]
  suggested: string
  reason: 'remembered' | 'named' | 'recent' | 'first'
}

async function fetchStartRepo(
  project: string,
  ticket: string,
): Promise<StartRepoHint> {
  const res = await apiFetch(
    `/api/v1/projects/${encodeURIComponent(project)}/start-repo?id=${encodeURIComponent(ticket)}`,
  )
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'start repo request failed'))
  }
  return res.json()
}

export function startRepoQueryOptions(project: string, ticket: string) {
  return queryOptions({
    queryKey: ['start-repo', project, ticket],
    queryFn: () => fetchStartRepo(project, ticket),
    enabled: project !== '',
    staleTime: 30_000,
  })
}

async function errorMessage(res: Response, fallback: string): Promise<string> {
  const detail = (await res.json().catch(() => null)) as {
    error?: string
  } | null
  return detail?.error ?? `${fallback}: ${res.status}`
}

export async function createProject(name: string): Promise<ProjectView> {
  const res = await apiFetch('/api/v1/projects', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'create project failed'))
  }
  return res.json()
}

export async function renameProject(id: string, name: string): Promise<ProjectView> {
  const res = await apiFetch(`/api/v1/projects/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'rename project failed'))
  }
  return res.json()
}

// ProjectTrackerKey is one key of a project's shared tracker. A secret carries
// no value, only whether one is stored.
export interface ProjectTrackerKey {
  key: string
  value?: string
  secret?: boolean
  set?: boolean
}

export interface ProjectTracker {
  project: string
  repos: string[]
  keys: ProjectTrackerKey[]
}

async function fetchProjectTracker(id: string): Promise<ProjectTracker> {
  const res = await apiFetch(
    `/api/v1/projects/${encodeURIComponent(id)}/tracker`,
  )
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'project tracker request failed'))
  }
  return res.json()
}

export function projectTrackerQueryOptions(id: string) {
  return queryOptions({
    queryKey: ['project-tracker', id],
    queryFn: () => fetchProjectTracker(id),
    enabled: id !== '',
  })
}

export const PROJECT_DEFAULT_KEYS = ['READY_LABEL', 'EPIC_FLOW']

export interface ProjectDefaults {
  readyLabel: string
  epicFlow: boolean
}

// A project holding no EPIC_FLOW reads as on, matching what its members resolve
// today from the shipped default — a fallback of off would show a state no repo
// is in and cost a round trip to actually turn the flow off.
export function projectDefaults(
  tracker: ProjectTracker | undefined,
): ProjectDefaults {
  const value = (key: string) =>
    tracker?.keys.find((k) => k.key === key)?.value ?? ''
  return {
    readyLabel: value('READY_LABEL'),
    epicFlow: value('EPIC_FLOW') !== '0',
  }
}

// projectDefaultsModified is the nav's dot: whether the project answers either
// field itself rather than leaving both to the defaults.
export function projectDefaultsModified(
  tracker: ProjectTracker | undefined,
): boolean {
  const keys = tracker?.keys ?? []
  return keys.some(
    (k) => PROJECT_DEFAULT_KEYS.includes(k.key) && (k.value ?? '') !== '',
  )
}

// Only the fields the user edited are written: a key left out of the request
// keeps whatever the project already stores, so saving one field never restates
// the other. A blank ready label is an edit like any other — it drops the
// project's answer and takes the key back off the members holding it on the
// project's behalf.
export function projectDefaultKeys(
  edited: Partial<ProjectDefaults>,
): Record<string, string> {
  const keys: Record<string, string> = {}
  if (edited.readyLabel !== undefined) {
    keys.READY_LABEL = edited.readyLabel.trim()
  }
  if (edited.epicFlow !== undefined) {
    keys.EPIC_FLOW = edited.epicFlow ? '1' : '0'
  }
  return keys
}

const PROJECT_DEFAULT_TERMS = [
  'project defaults',
  'ready label',
  'ready_label',
  'epic flow',
  'epic_flow',
]

export function matchesProjectDefaults(query: string): boolean {
  return matchesTerms(query, PROJECT_DEFAULT_TERMS)
}

// writeProjectTracker configures the tracker once for the whole project; the hub
// seeds the keys into every member repo's .trau.ini. A key left out — or a blank
// secret — keeps whatever is already stored.
export async function writeProjectTracker(
  id: string,
  keys: Record<string, string>,
): Promise<ProjectTracker> {
  const res = await apiFetch(
    `/api/v1/projects/${encodeURIComponent(id)}/tracker`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ keys }),
    },
  )
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'save project tracker failed'))
  }
  return res.json()
}

// ProjectRemovalRepo is one member a removal reached; reason is the hub's refusal
// on a member that stayed.
export interface ProjectRemovalRepo {
  name: string
  root: string
  reason?: string
}

// ProjectRemoval is the outcome of removing a project's repos. project is the
// pre-removal snapshot, so the caller can name what it acted on even once the group
// is gone, and blocked carries the members the hub kept: the removal is partial
// rather than all-or-nothing, so a success can still leave members behind.
export interface ProjectRemoval {
  project: ProjectView
  removed: ProjectRemovalRepo[]
  blocked: ProjectRemovalRepo[]
  project_deleted: boolean
}

// removeProjectRepos takes every member repo off the hub in one call, the same
// removal the per-repo Remove button makes: nothing on disk is touched, and running
// trau in a member again brings it back.
export async function removeProjectRepos(id: string): Promise<ProjectRemoval> {
  const res = await apiFetch(
    `/api/v1/projects/${encodeURIComponent(id)}?forget=1`,
    { method: 'DELETE' },
  )
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'remove project failed'))
  }
  return res.json()
}

// BlockedMember is a member the hub would keep, with the reason the row already
// shows for it.
export interface BlockedMember {
  name: string
  reason: string
}

// RemovalPlan is what the confirm dialog names before the click: the members a
// project-wide removal takes, and the ones it would leave with the reason each is
// kept for. It reads the repo list the page already holds, so the hub stays the
// authority — a member removable here can still come back blocked in the outcome.
export interface RemovalPlan {
  removing: string[]
  blocked: BlockedMember[]
}

export function removalPlan(repos: readonly RepoView[]): RemovalPlan {
  const plan: RemovalPlan = { removing: [], blocked: [] }
  for (const repo of repos) {
    const reason = removeBlocked(repo)
    if (reason === null) {
      plan.removing.push(repo.name)
      continue
    }
    plan.blocked.push({ name: repo.name, reason })
  }
  return plan
}

// removalSummary is what a removal's toast says: the count that left when the folder
// cleared, or which members stayed and why when the hub kept some.
export function removalSummary({
  project,
  removed,
  blocked,
}: ProjectRemoval): string {
  if (blocked.length === 0) {
    const repos = removed.length === 1 ? 'repo' : 'repos'
    return `${removed.length} ${repos} removed from ${project.name}`
  }
  const stayed = blocked
    .map((repo) => `${repo.name} stayed (${repo.reason})`)
    .join(', ')
  const total = removed.length + blocked.length
  return `${removed.length} of ${total} removed from ${project.name} — ${stayed}`
}

// addProjectRepo moves an already-registered repo — named or addressed by root —
// into the project, out of whichever project held it.
export async function addProjectRepo(
  id: string,
  repo: string,
): Promise<ProjectView> {
  const res = await apiFetch(
    `/api/v1/projects/${encodeURIComponent(id)}/repos`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ repo }),
    },
  )
  if (!res.ok) {
    throw new Error(await errorMessage(res, 'add repo failed'))
  }
  return res.json()
}
